src/api/ml/CMakeLists.txt

```diff
@@ -0,0 +1,320 @@
+  COMMAND "${OCAMLFIND}" "ocamlc" ${z3ml_common_flags}
+          "-o" "${z3ml_bin}/z3native_stubs.o"
+          "-I" "${z3ml_src}"
+          "-I" "${PROJECT_SOURCE_DIR}/src/api"
+          "-I" "${ocaml_stub_lib_path}"
+          "-c" "${z3ml_bin}/z3native_stubs.c"
+  DEPENDS ${z3ml_bin}/z3native_stubs.c
+  COMMENT "Building z3native_stubs.o"
+  VERBATIM)
+
+# z3enum.ml depends on nothing
+add_custom_command(
+  OUTPUT ${z3ml_bin}/z3enums.mli
+         ${z3ml_bin}/z3enums.cmi
+         ${z3ml_bin}/z3enums.cmo
+         ${z3ml_bin}/z3enums.cmx
+  COMMAND "${OCAMLFIND}" "ocamlc"   ${z3ml_common_flags}
+         "-i" 
+         "-c" "${z3ml_bin}/z3enums.ml"
+         ">" "${z3ml_bin}/z3enums.mli"
+  COMMAND "${OCAMLFIND}" "ocamlc"   ${z3ml_common_flags}
+         "-c" "${z3ml_bin}/z3enums.mli"  
+  COMMAND "${OCAMLFIND}" "ocamlc"   ${z3ml_common_flags}
+         "-c" "${z3ml_bin}/z3enums.ml"
+  COMMAND "${OCAMLFIND}" "ocamlopt" ${z3ml_common_flags}
+          "-c" "${z3ml_bin}/z3enums.ml"
+  DEPENDS ${z3ml_bin}/z3enums.ml
+  COMMENT "Building z3enums.{mli,cmi,cmo,cmx}"
+  VERBATIM)
+
+# z3native.ml depends on z3enums
+add_custom_command(
+  OUTPUT ${z3ml_bin}/z3native.mli
+         ${z3ml_bin}/z3native.cmi
+         ${z3ml_bin}/z3native.cmo
+         ${z3ml_bin}/z3native.cmx
+  COMMAND "${OCAMLFIND}" "ocamlc"   ${z3ml_common_flags}
+         "-i" 
+         "-c" "${z3ml_bin}/z3native.ml"
+         ">" "${z3ml_bin}/z3native.mli"
+  COMMAND "${OCAMLFIND}" "ocamlc"   ${z3ml_common_flags}
+         "-c" "${z3ml_bin}/z3native.mli"  
+  COMMAND "${OCAMLFIND}" "ocamlc"   ${z3ml_common_flags}
+          "-c" "${z3ml_bin}/z3native.ml"
+  COMMAND "${OCAMLFIND}" "ocamlopt" ${z3ml_common_flags}
+          "-c" "${z3ml_bin}/z3native.ml"
+  DEPENDS ${z3ml_bin}/z3enums.cmo
+          ${z3ml_bin}/z3native.ml
+  COMMENT "Building z3native.{mli,cmi,cmo,cmx}"
+  VERBATIM)
+
+# z3.ml depends on z3enums and z3native
+add_custom_command(
+  OUTPUT ${z3ml_bin}/z3.cmi
+         ${z3ml_bin}/z3.cmo
+         ${z3ml_bin}/z3.cmx
+  COMMAND "${OCAMLFIND}" "ocamlc"   ${z3ml_common_flags}
+          "-c" "${z3ml_bin}/z3.mli"
+  COMMAND "${OCAMLFIND}" "ocamlc"   ${z3ml_common_flags}
+          "-c" "${z3ml_bin}/z3.ml"
+  COMMAND "${OCAMLFIND}" "ocamlopt" ${z3ml_common_flags}
+          "-c" "${z3ml_bin}/z3.ml" 
+  DEPENDS ${z3ml_bin}/z3enums.cmo
+          ${z3ml_bin}/z3native.cmo
+          ${z3ml_bin}/z3.ml
+          ${z3ml_bin}/z3.mli
+  COMMENT "Building z3.cmo"
+  VERBATIM)
+
+# making ocaml stublibs  
+execute_process(
+  COMMAND ${OCAMLFIND} printconf destdir
+  OUTPUT_VARIABLE ocaml_destdir_path
+  OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+set(ocaml_stublibs_path "${ocaml_destdir_path}/stublibs")
+
+set(c_lib_deps "-lz3" "-lstdc++" "-lpthread")
+if (Z3_USE_LIB_GMP)
+  list(APPEND c_lib_deps "-lgmp")
+endif()
+
+if( APPLE )
+  set(ocaml_rpath "@executable_path/../libz3${so_ext}")
+elseif( UNIX )
+  set(ocaml_rpath "\\$ORIGIN/../libz3${so_ext}")
```

Comment:
At this moment, building on Windows/Cygwin is beyond my knowledge and I think even the support for OCaml on Windows is under heavy development. I may prefer to at least postpone it until the rest is done.

p.s. I notice there is some good news in the OCaml community.

Reply:
Yeah, it's always been a bit neglected, but there are users, so we should make sure it works. BTW, the Opam people are also adding a Windows CI pipeline at the moment, so it might be required for Opam packages soon.